### PR TITLE
Update prev_node_idx in loop when removing entry.

### PR DESCRIPTION
**Pull request by toregge**
Use next_node_idx instead of next.
Use first_used instead of used_gen or usedGen.

@baldersheim : please review

### CODE REVIEW

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+FixedSizeHashMapHeld::FixedSizeHashMapHeld(size_t size, std::unique_ptr<const FixedSizeHashMap> data)
+    : GenerationHeldBase(size),
+      _data(std::move(data))
+{
+}
+
+FixedSizeHashMapHeld::~FixedSizeHashMapHeld() = default;
+
+struct DataStoreFixedSizeHashTest : public ::testing::Test
```
Comment (geirst) [inline vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp:45]: Just copying the tests from simple_hash_map_test.cpp (with slight modifications) doesn't give any extra value in my opinion. The tests for FixedSizeHashMap should be on a lower level. No need for tests using multiple writers and readers that tests with a lot of data. That is already handled in simple_hash_map_test.cpp. We should write test cases the triggers the particular edge cases instead. For instance ensuring that the linked list is updated correctly when removing and ensuring that free nodes are re-used etc (perhaps we need to expose _free_count for this). I think only smoke_test and lookups_works_after_insert_and_remove is on the right abstraction level, the rest can go. Growing the hash map on the fly seems to just complicate things as well.

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    std::atomic<long> _done_read_work;
+    std::atomic<long> _found_count;
+    std::atomic<int> _stop_read;
+    size_t _modulo_limit;
+    bool _report_work;
+
+    DataStoreFixedSizeHashTest();
+    ~DataStoreFixedSizeHashTest();
+    void commit();
+    void grow();
```
Comment (geirst) [inline vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp:68]: Consider writing tests that doesn't need growing. Let's keep these tests as simple as possible.

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        commit();
+    }
+    auto hash_map = _hash_map.load(std::memory_order_acquire);
+    for (auto &kv : expected) {
+        MyCompare comp(_store, kv.first);
+        EXPECT_EQ(kv.second, hash_map->find(comp, EntryRef()) != nullptr);
+    }
+}
+
+TEST_F(DataStoreFixedSizeHashTest, single_threaded_reader_without_updates)
```
Comment (geirst) [inline vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp:286]: This is mainly covered in simple_hash_map_test.cpp, consider removing.

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+{
+    _report_work = true;
+    write_work(10);
+    _stop_read = 0;
+    read_work(10);
+}
+
+TEST_F(DataStoreFixedSizeHashTest, single_threaded_reader_during_updates)
```
Comment (geirst) [inline vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp:294]: This is mainly covered in simple_hash_map_test.cpp, consider removing.

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+{
+    uint32_t cnt = 1000000;
+    _report_work = true;
+    _writer.execute(makeLambdaTask([this, cnt]() { write_work(cnt); }));
+    _readers.execute(makeLambdaTask([this]() { read_work(); }));
+}
+
+TEST_F(DataStoreFixedSizeHashTest, multi_threaded_reader_during_updates)
```
Comment (geirst) [inline vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp:302]: This is mainly covered in simple_hash_map_test.cpp, consider removing.

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    insert(4);
+    EXPECT_EQ(3, size());
+    remove(3);
+    EXPECT_EQ(3, size());
+    remove(5);
+    EXPECT_EQ(2, size());
+    commit();
+    MyCompare comp3(_store, 3);
+    auto hash_map = _hash_map.load(std::memory_order_acquire);
+    auto result3 = hash_map->find(comp3, EntryRef());
```
Comment (geirst) [inline vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp:253]: Consider creating helper function(s) to verify find().

**Review by baldersheim — CHANGES_REQUESTED** [review]
I agree with most @geirst comments, but I still think that threading tests belong at this level. They are more important here than at the simple_hash_map level. But growing tests belong up there.

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp (hunk 2 above)
Comment (toregge) [reply vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp]: Growing removed from test.

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp (hunk 4 above)
Comment (toregge) [reply vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp]: Removed.

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp (hunk 3 above)
Comment (toregge) [reply vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp]: Removed.

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp (hunk 5 above)
Comment (toregge) [reply vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp]: Removed.

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp (hunk 6 above)
Comment (toregge) [reply vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp]: Done.

File: vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp (hunk 1 above)
Comment (toregge) [reply vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp:45]: Threaded tests have been removed.

**Review by geirst — APPROVED** [review]
👍  @baldersheim please have a look as well

**Review by baldersheim — APPROVED** [review]
